homeassistant/components/zerproc/light.py

```diff
@@ -0,0 +1,198 @@
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+SUPPORT_ZERPROC = SUPPORT_BRIGHTNESS | SUPPORT_COLOR
+
+DISCOVERY_INTERVAL = timedelta(seconds=60)
+
+PARALLEL_UPDATES = 0
+
+
+def connect_lights(lights: List[pyzerproc.Light]) -> List[pyzerproc.Light]:
+    """Attempt to connect to lights, and return the connected lights."""
+    connected = []
+    for light in lights:
+        try:
+            light.connect(auto_reconnect=True)
+            connected.append(light)
+        except pyzerproc.ZerprocException:
+            _LOGGER.debug("Unable to connect to '%s'", light.address, exc_info=True)
+
+    return connected
+
+
+def discover_entities(hass: HomeAssistant) -> List[Entity]:
+    """Attempt to discover new lights."""
+    lights = pyzerproc.discover()
+
+    # Filter out already discovered lights
+    new_lights = [
+        light
+        for light in lights
+        if light.address not in hass.data[DOMAIN]["light_entities"]
+    ]
+
+    entities = []
+    for light in connect_lights(new_lights):
+        # Double-check the light hasn't been added in another thread
+        if light.address not in hass.data[DOMAIN]["light_entities"]:
+            entity = ZerprocLight(hass, light)
+            hass.data[DOMAIN]["light_entities"][light.address] = entity
+            entities.append(ZerprocLight(hass, light))
+
+    return entities
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType,
+    config_entry: ConfigEntry,
+    async_add_entities: Callable[[List[Entity], bool], None],
+) -> None:
+    """Set up Abode light devices."""
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+    if "light_entities" not in hass.data[DOMAIN]:
+        hass.data[DOMAIN]["light_entities"] = {}
+
+    warned = False
+
+    async def discover(*args):
+        """Wrap discovery to include params."""
+        nonlocal warned
+        try:
+            entities = await hass.async_add_executor_job(discover_entities, hass)
+            async_add_entities(entities, update_before_add=True)
+            warned = False
+        except pyzerproc.ZerprocException:
+            if warned is False:
+                _LOGGER.warning("Error discovering Zerproc lights", exc_info=True)
+                warned = True
+
+    # Initial discovery
+    hass.async_add_job(discover)
+
+    # Perform recurring discovery of new devices
+    async_track_time_interval(hass, discover, DISCOVERY_INTERVAL)
+
+
+class ZerprocLight(Light):
+    """Representation of an Zerproc Light."""
+
+    def __init__(self, hass, light):
+        """Initialize a Zerproc light."""
+        self._light = light
+        self._name = None
+        self._is_on = None
+        self._hs_color = None
+        self._brightness = None
+        self._available = True
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_hass_shutdown)
```

Comment:
You will want to undo this listener if the entity is removed earlier. You probably also want to add `async_will_remove_from_hass` that will call `self._light.disconnect()`

```suggestion
        self.async_on_remove(
            self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_hass_shutdown)
        )
```